### PR TITLE
Prefix seek (and our own rocksdb build)

### PR DESCRIPTION
**Pull request by edmundnoble**
Depends on https://github.com/kadena-io/chainweb-storage/pull/14

### CODE REVIEW

**Comment by edmundnoble** [comment]
Benchmarks using [hyperfine]():
```
# master
edmundnoble@mx:~/kadena/infrastructure/chainweb/bootstrap-nodes/testnet/us2-testnet-chainweb-com
$ hyperfine -m 100 --warmup 20 "curl -sk 'https://us1.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/header' -H 'content-type:application/json' -H 'accept:application/json;blockheader-encoding=object' -v 2>/dev/null | jq '[.items[].payloadHash]' | curl -sk 'https://us1.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/payload/batch' -d @- -H 'content-type:application/json' > /dev/null 2>&1"
Benchmark 1: curl -sk 'https://us1.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/header' -H 'content-type:application/json' -H 'accept:application/json;blockheader-encoding=object' -v 2>/dev/null | jq '[.items[].payloadHash]' | curl -sk 'https://us1.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/payload/batch' -d @- -H 'content-type:application/json' > /dev/null 2>&1
  Time (mean ± σ):     529.3 ms ±  18.7 ms    [User: 94.3 ms, System: 20.9 ms]
  Range (min … max):   482.0 ms … 582.2 ms    100 runs
```
```
# ae84544 (this PR)
edmundnoble@mx:~/kadena/infrastructure/chainweb/bootstrap-nodes/testnet/us2-testnet-chainweb-com
$ hyperfine -m 100 --warmup 20 "curl -sk 'https://us2.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/header' -H 'content-type:application/json' -H 'accept:application/json;blockheader-encoding=object' -v 2>/dev/null | jq '[.items[].payloadHash]' | curl -sk 'https://us2.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/payload/batch' -d @- -H 'content-type:application/json' > /dev/null 2>&1"
Benchmark 1: curl -sk 'https://us2.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/header' -H 'content-type:application/json' -H 'accept:application/json;blockheader-encoding=object' -v 2>/dev/null | jq '[.items[].payloadHash]' | curl -sk 'https://us2.testnet.chainweb.com/chainweb/0.0/testnet04/chain/0/payload/batch' -d @- -H 'content-type:application/json' > /dev/null 2>&1
  Time (mean ± σ):     514.6 ms ±  50.1 ms    [User: 78.0 ms, System: 17.3 ms]
  Range (min … max):   447.8 ms … 747.0 ms    100 runs
```

`-m 100` means 100 iterations, `--warmup 20` is 20 warmup iterations.

These are the most stable numbers I got, there were usually outliers. It *looked* like there were more on master, but master was on us1 which might be used more often, and they weren't consistent enough to include in the benchmark.

**Comment by edmundnoble** [comment]
Also a note, the update of `cabal-version` is *required* for the correct behavior of `extra-bundled-libraries`.

File: .github/actions/install-chainweb-from-artifacts/action.yml [hunk 1]
```diff
@@ -17,10 +17,10 @@ runs:
       run: |
         case "${{ inputs.os }}" in
         ubuntu*)
-          sudo apt-get install -y librocksdb-dev z3
+          sudo apt-get install -y z3 build-essential libsnappy-dev libgflags-dev libbz2-dev
           ;;
         mac*)
-          brew update && brew install rocksdb z3 || true
+          brew update && brew install gflags rocksdb llvm snappy bzip2 z3 gnu-tar || true
```
Comment (larskuhtz) [inline .github/actions/install-chainweb-from-artifacts/action.yml:23]: remove rocksdb?

File: .github/workflows/applications.yml [hunk 2]
```diff
@@ -567,7 +567,7 @@ jobs:
         LABEL com.chainweb.docker.image.optimization="$OPT_LEVEL"
         LABEL com.chainweb.docker.image.debug-info="$DEBUG_INFO"
         LABEL com.chainweb.docker.image.revision="$GIT_SHA_SHORT"
-        RUN apt-get update && apt-get install -y librocksdb-dev ca-certificates locales && rm -rf /var/lib/apt/lists/* && locale-gen en_US.UTF-8 && update-locale LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
+        RUN apt-get update && apt-get install -y libgflags-dev libsnappy-dev z3 libbz2-dev ca-certificates locales && rm -rf /var/lib/apt/lists/* && locale-gen en_US.UTF-8 && update-locale LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
```
Comment (larskuhtz) [inline .github/workflows/applications.yml:570]: Why are the development libraries needed to run chainweb-node? I would expect the following to work, assuming that rocksdb is linked statically

```suggestion
        RUN apt-get update && apt-get install -y ca-certificates locales && rm -rf /var/lib/apt/lists/* && locale-gen en_US.UTF-8 && update-locale LC_ALL=en_US.UTF-8 LANG=en_US.UTF-8
```

File: .github/workflows/macos.yaml [hunk 3]
```diff
@@ -33,7 +33,7 @@ jobs:
     - name: Install non-Haskell dependencies (macOS)
       if: contains(matrix.os, 'mac')
       run: |
-        brew update && brew install rocksdb z3 gnu-tar || true
+        brew update && brew install llvm z3 gnu-tar rocksdb || true
```
Comment (larskuhtz) [inline .github/workflows/macos.yaml:36]: Why rocksdb, aren't we building it from scratch?

```suggestion
        brew update && brew install llvm z3 gnu-tar || true
```

File: .github/actions/install-chainweb-from-artifacts/action.yml (hunk 1 above)
Comment (larskuhtz) [inline .github/actions/install-chainweb-from-artifacts/action.yml:23]: This action installs chainweb-node from the artifacts. Why are the build tools and development libraries needed?

```suggestion
          brew update && brew install gflags gnu-tar || true
```

File: cabal.project [hunk 4]
```diff
@@ -36,6 +36,18 @@ package hashes
...
+      /opt/local/include
+      /usr/local/opt/gflags/include
+      /usr/local/opt/rocksdb/include
+      /usr/local/include
+    extra-lib-dirs:
+      /opt/local/lib
+      /usr/local/opt/gflags/lib
+      /usr/local/opt/rocksdb/lib
+      /usr/local/lib
+
```
Comment (larskuhtz) [inline cabal.project:50]: shouldn't rocksdb be installed within the scope of the `rocksdb-haskell` package within the `~/.cabal` directory?

File: cabal.project [hunk 5]
```diff
@@ -36,6 +36,16 @@ package hashes
...
+package rocksdb-haskell
+    extra-include-dirs:
+      /opt/local/include
+      /usr/local/opt/gflags/include
+      /usr/local/include
+    extra-lib-dirs:
+      /opt/local/lib
+      /usr/local/opt/gflags/lib
+      /usr/local/lib
+
```
Comment (larskuhtz) [inline cabal.project:48]: Why is this still needed? I think it can be removed now
```suggestion
```

File: chainweb.cabal [hunk 6]
```diff
@@ -1,4 +1,4 @@
-cabal-version: 2.4
+cabal-version: 3.0
```
Comment (larskuhtz) [inline chainweb.cabal:1]: 👍

File: .github/actions/install-chainweb-from-artifacts/action.yml [hunk 7]
```diff
@@ -17,10 +17,10 @@ runs:
       run: |
         case "${{ inputs.os }}" in
         ubuntu*)
-          sudo apt-get install -y librocksdb-dev z3
+          sudo apt-get install -y z3 build-essential
```
Comment (larskuhtz) [inline .github/actions/install-chainweb-from-artifacts/action.yml:20]: I think, z3 can be dropped here as well, but not 100% sure.
```suggestion
          sudo apt-get install -y build-essential 
```

File: .github/workflows/applications.yml [hunk 8]
```diff
@@ -226,8 +226,7 @@ jobs:
         cabal --version
     - name: Install non-Haskell dependencies (ubuntu)
       if: contains(matrix.os, 'ubuntu')
-      run: sudo apt-get install -y librocksdb-dev z3
-
+      run: sudo apt-get install -y liblz4-dev libzstd-dev libgflags-dev libsnappy-dev z3 libbz2-dev
```
Comment (larskuhtz) [inline .github/workflows/applications.yml:229]:
```suggestion
      run: sudo apt-get install -y liblz4-dev libzstd-dev libgflags-dev libsnappy-dev libbz2-dev
```

File: .github/workflows/macos.yaml [hunk 9]
```diff
@@ -33,7 +33,7 @@ jobs:
     - name: Install non-Haskell dependencies (macOS)
       if: contains(matrix.os, 'mac')
       run: |
-        brew update && brew install rocksdb z3 gnu-tar || true
+        brew update && brew install gflags llvm z3 gnu-tar snappy zstd lz4 || true
```
Comment (larskuhtz) [inline .github/workflows/macos.yaml:36]:
```suggestion
        brew update && brew install gflags llvm gnu-tar snappy zstd lz4 || true
```

File: bench/Chainweb/Pact/Backend/ForkingBench.hs [hunk 10]
```diff
@@ -98,6 +98,7 @@ import Chainweb.Version.Utils
 
 import Data.CAS.HashMap hiding (toList)
 import Data.CAS.RocksDB
+import qualified Database.RocksDB.Internal as R
```
Comment (larskuhtz) [inline bench/Chainweb/Pact/Backend/ForkingBench.hs:101]: Is this required? I would expect the `R` qualifier to be used somewhere in the file.

File: chainweb.cabal [hunk 11]
```diff
@@ -534,6 +534,7 @@ test-suite chainweb-tests
         , random >= 1.2
         , resource-pool >= 0.3
         , retry >= 0.7
+        , rocksdb-haskell-kadena
```
Comment (larskuhtz) [inline chainweb.cabal:537]: For idiomatic consistency with the rest of this cabal file

```suggestion
        , rocksdb-haskell-kadena >= 1.1.0
```

File: chainweb.cabal [hunk 12]
```diff
@@ -748,6 +749,7 @@ benchmark bench
         , merkle-log
         , mtl >= 2.2
         , pact
+        , rocksdb-haskell-kadena
```
Comment (larskuhtz) [inline chainweb.cabal:752]:
```suggestion
        , rocksdb-haskell-kadena >= 1.1.0
```

File: test/Chainweb/Test/TreeDB.hs [hunk 13]
```diff
@@ -139,7 +139,7 @@ reinsertion_prop
 reinsertion_prop f (SparseTree t0) = ioProperty . withTreeDb f t $ \db insert -> do
     insert db (toList t)
     l <- entries db Nothing Nothing Nothing Nothing P.length_
-    pure $ l == length t
+    return $ l === length t
```
Comment (larskuhtz) [inline test/Chainweb/Test/TreeDB.hs:142]: 👍

File: test/Chainweb/Test/Utils.hs [hunk 14]
```diff
@@ -280,24 +281,29 @@ testRocksDb
...
       dir <- createTempDirectory sysdir "chainweb-rocksdb-tmp"
-      rocks <- openRocksDb dir
-      return (dir, rocks)
-    destroy (dir, rocks) = do
+      opts@(R.Options' opts_ptr _ _) <- R.mkOpts modernDefaultOptions
+      rocks <- openRocksDb dir opts_ptr
+      return (dir, rocks, opts)
+    destroy (dir, rocks, opts) = do
         closeRocksDb rocks
+          `catchAllSynchronous` (const $ return ())
```
Comment (larskuhtz) [inline test/Chainweb/Test/Utils.hs:299]: Could this possibly hide exception that we'd like to catch during testing? If there are expected exceptions, maybe we could be more specific in what we catch?

File: test/Chainweb/Test/Utils.hs [hunk 15]
```diff
@@ -280,24 +281,29 @@ testRocksDb
...
+        R.freeOpts opts
+          `catchAllSynchronous` (const $ return ())
         destroyRocksDb dir
+          `catchAllSynchronous` (const $ return ())
         removeDirectoryRecursive dir
           `catchAllSynchronous` (const $ return ())
-    wrap ioact = let io' = snd <$> ioact in m io'
+    wrap ioact = let io' = view _2 <$> ioact in m io'
```
Comment (larskuhtz) [inline test/Chainweb/Test/Utils.hs:306]: What is the benefit of this change? Just curious...

**Review by larskuhtz — CHANGES_REQUESTED** [review]

File: test/Chainweb/Test/Utils.hs (hunk 15 above)
Comment (edmundnoble) [reply test/Chainweb/Test/Utils.hs:306]: IIRC `ioact` is now a triple, not a pair.

File: test/Chainweb/Test/Utils.hs (hunk 14 above)
Comment (edmundnoble) [reply test/Chainweb/Test/Utils.hs:299]: Replacing with `finally`.

**Review by larskuhtz — APPROVED** [review]

**Pull request merged**